posthog/models/tagged_item.py

```diff
@@ -2,23 +2,36 @@
 
 from django.core.exceptions import ValidationError
 from django.db import models
-from django.db.models import Q
+from django.db.models import Q, UniqueConstraint
 
 from posthog.models.utils import UUIDModel
 
 RELATED_OBJECTS = ("dashboard", "insight", "event_definition", "property_definition", "action")
 
 
+# Checks that exactly one object field is populated
 def build_check():
     built_check_list: List[Union[Q, Q]] = []
-    # Only one object field can be populated
-    for o_field in RELATED_OBJECTS:
+    for field in RELATED_OBJECTS:
         built_check_list.append(
-            Q(*[(f"{_o_field}__isnull", _o_field != o_field) for _o_field in RELATED_OBJECTS], _connector="AND")
+            Q(*[(f"{other_field}__isnull", other_field != field) for other_field in RELATED_OBJECTS], _connector="AND")
         )
     return Q(*built_check_list, _connector="OR")
 
 
+# Enforces uniqueness on tag_{object_field}. All permutations of null columns must be explicit as Postgres ignores
+# uniqueness across null columns.
+def build_partial_uniqueness_constraint(field: str):
+    return UniqueConstraint(
+        fields=["tag", field],
+        name=f"unique_{field}_tagged_item",
+        condition=Q(
+            *[(f"{other_field}__isnull", True) for other_field in RELATED_OBJECTS if other_field != field],
```

Comment:
I'm still unsure why this needs to know about other fields and not just check for nullness of the field in question